src/SignalR/server/Core/src/Internal/AsyncEnumerableAdapters.cs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.SignalR.Internal
+{
+    // True-internal because this is a weird and tricky class to use :)
+    internal static class AsyncEnumerableAdapters
+    {
+        public static IAsyncEnumerable<object> MakeCancelableAsyncEnumerable<T>(IAsyncEnumerable<T> asyncEnumerable, CancellationToken cancellationToken = default)
+        {
+            return new CancelableAsyncEnumerable<T>(asyncEnumerable, cancellationToken);
+        }
+
+        public static IAsyncEnumerable<object> MakeCancelableAsyncEnumerableFromChannel<T>(ChannelReader<T> channel, CancellationToken cancellationToken = default)
+        {
+            return MakeCancelableAsyncEnumerable(channel.ReadAllAsync(), cancellationToken);
+        }
+
+        /// <summary>Converts an IAsyncEnumerable of T to an IAsyncEnumerable of object.</summary>
+        private class CancelableAsyncEnumerable<T> : IAsyncEnumerable<object>
+        {
+            private readonly IAsyncEnumerable<T> _asyncEnumerable;
+            private readonly CancellationToken _cancellationToken;
+
+            public CancelableAsyncEnumerable(IAsyncEnumerable<T> asyncEnumerable, CancellationToken cancellationToken)
+            {
+                _asyncEnumerable = asyncEnumerable;
+                _cancellationToken = cancellationToken;
+            }
+
+            public IAsyncEnumerator<object> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+            {
+                // Assume that this will be iterated through with await foreach which always passes a default token.
+                // Instead use the token from the ctor.
+                Debug.Assert(cancellationToken == default);
```

Comment:
REVIEW: The reason we have to wrap IAsyncEnumerators with non-value-type generic parameters instead of just casting is because we iterate using `await foreach` which doesn't support flowing a CancellationToken through GetAsyncEnumerator. If we manually iterated over the enumerator, we could save ourselves wrapping in most cases.

So should we manually iterate instead of wrapping the GetAsyncEnumerator call?

@davidfowl

Reply:
We are the only consumer of the `AsyncEnumerable<object>` btw

Reply:
So we could loop over `MoveNextAsync()` manually and avoid the `BoxedAsyncEnumerator`? That sounds nice :D

Reply:
Yes we should optimize

Reply:
You could have said something before I merged. It *was* a while loop that manually called GetAsyncEnumerator/MoveNextAsync until you asked me to change it to `await foreach` @davidfowl 😆

I figured didn't care about the allocation since a bunch of other allocations happen in this code path anyway. I created #7960 to track this with an XS tag. If we want to change this, it's not difficult.

Reply:
I care, lets fix it in preview4. Assign it to yourself 😄

Reply:
I'm glad you care 😉